File: 2020/day_6/day_6.py
"""Puzzle Day 6"""
from pathlib import Path

import pytest


def get_common_in_group(group):
    results = set(group[0])
    for passenger in group[1:]:
        results.intersection_update(passenger)
    return results


if __name__ == "__main__":
    with open(Path(__file__).parent.joinpath("input.txt")) as puzzle_input:
        passenger_groups = [
            [passengers for passengers in group.split("\n") if passengers]
            for group in puzzle_input.read().split("\n\n")
        ]

    # Spliting with "\n\n" is going to add an empty element at the end.
    anyone_answered_yes = [
        len(set("".join(passengers))) for passengers in passenger_groups
    ]

    print(
        f"Sum of yes for anyone in passenger groups: {sum(anyone_answered_yes)}"
    )  # 6259

    group_answered_yes = list()
    for group in passenger_groups:
        group_answered_yes.append(len(get_common_in_group(group)))

    print(
        f"Sum of yes for everyone in passenger groups: {sum(group_answered_yes)}"
    )  # 3178


@pytest.mark.parametrize(
    "test,expected",
    [
        (["abc"], 3),
        (["a", "b", "c"], 0),
        (["ab", "ac"], 1),
        (["a", "a", "a", "a"], 1),
        (["b"], 1),
    ],
)
def test_get_common_in_group(test, expected):
    assert len(get_common_in_group(test)) == expected

File: 2020/day_6/day_6.cpp
// std::set<std::string> a;
#include <algorithm>
#include <iostream>
#include <iterator>
#include <fstream>
#include <numeric>
#include <vector>
#include <set>

void read_in_customs_forms( std::vector<std::vector<std::set<char>>> &passenger_groups, const std::string filename );

int main()
{
    std::vector<std::vector<std::set<char>>> passenger_groups;
    read_in_customs_forms( passenger_groups, "input.txt" );

    int anyone_who_said_yes {0};
    for( auto group : passenger_groups )
    {
        std::set<char> yes_in_group {};
        for( auto passenger : group )
        {
            yes_in_group.insert( passenger.begin(), passenger.end() );
        }
        anyone_who_said_yes += yes_in_group.size();
    }
    std::cout << "Sum of yes for anyone in passenger groups: " << anyone_who_said_yes << std::endl; // 6259

    int everyone_said_yes {0};

    for( auto group : passenger_groups )
    {
        std::set<char> result {group.at( 0 )};
        for( int index = 1; index < group.size(); index++ )
        {
            std::set<char> intersection;
            std::set_intersection( result.begin(), result.end(), group.at( index ).begin(), group.at( index ).end(),  std::inserter( intersection, intersection.begin() ) );
            result = intersection;
        }
        everyone_said_yes += result.size();
    }
    std::cout << "Sum of yes for everyone in passenger groups: " << everyone_said_yes << std::endl; // 3178

}

void read_in_customs_forms( std::vector<std::vector<std::set<char>>> &passenger_groups, const std::string filename )
{
    std::ifstream file( filename );
    std::string line;
    std::vector<std::set<char>> group;
    std::set<char> passenger;
    while( std::getline( file, line ) )
    {
        if( line.empty() )
        {
            passenger_groups.push_back( group );
            group.clear();
        }
        else
        {
            for( auto character : line )
            {
                passenger.insert( character );
            }
            group.push_back( passenger );
            passenger.clear();
        }

    }
    passenger_groups.push_back( group );
}
